### ISSUE TITLE
Uploading duplicates

### ISSUE THREAD

**Issue by a575606** [jaymoulin/google-music-manager#15]
I was excited to find your docker container because I keep my music on a NAS so I'd rather have the nas upload to google music as it's running all the time anyway. The problem is, most of my library is already uploaded and it seems the program doesn't check for duplicates before uploading? I started getting a lot of duplicate tracks so I stopped the container. How can that be avoided?

**Comment by jaymoulin** [comment]
TL;DR; You can't

Google Music Uploader is already checking for duplicates by sending (and i's important to understand that it's REQUIRED by Google) a sample BEFORE uploading a music.
The problem here is, Google sometimes gives false positive (or negative) and you can't do anything about that. The only solution I suggested for this but seems not to accepted by the community was deleting the local file after successful upload to Google (and it's the solution I personally use everyday)

(I suggested to keep a list of all already uploaded music as well but this require to reupload ALL your music library and still handle false positive/negative and you can't be able to move/rename any of your file)

BUT

I understand most of you want to keep their music library locally and I asked the previous user how would he do it since many solution with cons can be implemented (https://github.com/jaymoulin/google-music-manager/issues/12). No answer.

But, hey, this is OpenSource you know, so, I can accept PR ;)

By the way, I still accept donations if you find this program useful <3

**Comment by a575606** [comment]
Hmm, a little trigger happy on the close issue button no? Couldn't it be implemented with a local sqllite db of which tracks have already been uploaded?

**Comment by jaymoulin** [comment]
>(I suggested to keep a list of all already uploaded music as well but this require to reupload ALL your music library and still handle false positive/negative and you can't be able to move/rename any of your file)

Did you read my answer? This solution comes with a list of problem I'm not willing to deal with

**Comment by jaymoulin** [comment]
Anyway, up to you to send a PR and handle those issues ;)

**Comment by jaymoulin** [comment]
In fact, you're right. I'll leave this issue open for one month. If this issue doesn't gives anything, I'll close and lock it as I did for #12. I would be pleased to assign it to you so everyone knows somebody is working on it

**Comment by a575606** [comment]
Well, that was kind of a dick move if you ask me, but I took a look through your code anyway. Seems to me the simplest solution would be add a custom tag to mp3 files when they've been uploaded and check for that tag before uploading. Probably a quick if statement before line 91 in uploader_daemon.py would do the job. pypi.org has a python package for reading / writing mp3 tags. 

Fair enough if you don't want to spend the time on this issue. But assigning people who post bug reports the job of fixing the bugs / adding the enhancement? Seriously? I'd appreciate if you'd reverse that.

**Comment by jaymoulin** [comment]
I may not be the best human being in the world but:
- I offer my work freely.
- I don't go to people repository to insult them.
- I spend time answering to anyone even if they insult me and I'm not paid for that.
- I listen to people when I ask a question and they take time to answer me.
- I don't think telling people what to do will ever fix my own personal issues.

I don't deserve the "dick move" palm with this issue if you ask me but let's ignore this childish quarrel for the sake of the 100k+ downloader, will you?

--- 

Your idea is pretty good, but you have to be aware that it will **NOT** fix your issue.

Let me explain :
- If that specific tag doesn't exists, I'll have to check with Google to check if this file is already uploaded or not and thus have to deal with the false positive/negative issue.
- In case of a false positive, I will have to explain to each user who will come to the bug tracker that they have to delete a tag inside their MP3 (in case it's and MP3) to allow the program to try and upload it one more time.
- In case of a false negative, Google Music will still have duplicates.

On the other hand, this solution only works with MP3 files. Even if Google only allows MP3 files, I can not assume that users will have a full MP3 file music library. 

Do you have a better idea that matches with those constraints?

**Comment by a575606** [comment]
Well, I hardly think I was the one who started this, but yes let's drop the kids stuff. I certainly wasn't intending to visit your repository to insult you. Just hoping to see if there's a way to solve the problem. 

Understand your point. For me there are 2 options. Tag the files or maintain a database. And I think you could make it easy for users to understand by adding a 3rd volume, google library, in addition to upload and download. So on the initial run after initialization, just go through whatever is in that directory and either build a list or tag the files. Then whenever you scan that directory in future, any files not tagged or in the list get uploaded. 

Obviously it takes some thought to get the details of the algorithm right, but I think there's something in the idea and not a huge job to implement. What do you think?

**Comment by jaymoulin** [comment]
I don't think the "Google library" volume has any purpose.

I agree that the database is the best approach to keep track on what has already been tried or not without requiring Google to answer that.
Still, you have to understand that this solution will NOT resolve this issue, which is "How can I avoid having duplicates?". Are you OK with that?

Even if I ever implement a database based solution, this will only fix #12 and not #15 which is already implemented. If you agree with it, I'll close #15 and reopen #12 which seems to be more meaningful.

**Comment by a575606** [comment]
I suppose my idea with the google library volume is a way to pre-scan the music collection in order to build the database on first run. In my case, what I'm trying to achieve is, I have a music folder, all of which is already uploaded and working on google. I don't want to upload any of the existing tracks in there. What I want is for any new or modified music files I add to that folder from now onwards to be uploaded. 

So, I agree with your point it won't solve never having duplicates, but it will avoid potentially having my whole library duplicated. And if I add duplicate tracks in future, I'll understand where they came from. I ran your docker image for only a minute and out of 10 tracks it uploaded, 5 were duplicates, which would have been a real mess to clean up had I not caught it early. 

In terms of #12, no, I don't think that's the issue I'm describing. That has more to do with the efficiency of the disk scanning algorithm I think.

**Comment by a575606** [comment]
Further to my last comment, I installed Google Music manager again just to observe its behavior. Basically on first run, it scans the online collection and compares it to the local files before performing any uploads. It compiles a database in /users/user/AppData/local/Google/MusicManager/ServerDatabase.db to use for comparisons i assume. I don't have SqlLite on my pc but I'll do some research later to see what that contains. Perhaps that file could just be imported into your program and read from? Then kept updated afterwards?
